docs/cluster/networking.md

```diff
@@ -0,0 +1,142 @@
+the general mindset of Kubernetes.  Dynamic port allocation brings a lot of
+complications to the system - every application has to take ports as flags,
+the API servers have to know how to insert dynamic port numbers into
+configuration blocks, services have to know how to find each other, etc.  Rather
+than deal with this, Kubernetes takes a different approach.
+
+The fundamental requirements that Kubernetes makes about networking are:
+   * all containers can communicate with all other containers without NAT
+   * all nodes can communicate with all containers (and vice-versa) without NAT
+   * the IP that a container sees itself as is the same IP that others see it as
+
+What this means in practice is that you can not just take two computers
+running Docker and expect kubernetes to work.  You must ensure that the
+fundamental requirements are met.
+
+This model is not only less complex overall, but it is principally compatible
+with the desire for Kubernetes to enable low-friction migration from VMs.  If
+your job previously ran in a VM, your VM had an IP and could talk to other VMs
+in your project.  This is the same basic model.
+
+Until now this document has talked about containers.  In reality, Kubernetes
+applies IP addresses at the `Pod` scope - containers within a `Pod` share their
+network namespaces - including their IP address.  This means that containers
+within a `Pod` can all reach each other’s ports on `localhost`.  This does imply
+that containers within a `Pod` must coordinate port usage, but this is no
+different that processes in a VM.  We call this the "IP-per-pod" model.  It is
+possible to request host ports, but this is reduced to a very niche operation.
+
+## How to acheive this
+
+There are a number of ways that this network model can be implemented.  This
+document is not an exhaustive study of the various methods, but hopefully serves
+as an introduction to various technologies and serves as a jumping-off point.
+If some techniques become vastly preferable to others, we might detail them more
+here.
+
+### Google Compute Engine
+
+For the Google Compute Engine cluster configuration scripts, we use [advanced
+routing](https://developers.google.com/compute/docs/networking#routing) to
+assign each VM a subnet (default is /24 - 254 IPs).  Any traffic bound for that
+subnet will be routed directly to the VM by the GCE network fabric.  This is in
+addition to the "main" IP address assigned to the VM, which is NAT'ed for
+outbound internet access.  A linux bridge (called `cbr0`) is configured to exist
+on that subnet, and is passed to docker's `--bridge` flag.
+
+We start Docker with:
+
+```
+    DOCKER_OPTS="--bridge cbr0 --iptables=false --ip-masq=false"
+```
+
+We set up this bridge on each node with SaltStack, in
+[container_bridge.py](../cluster/saltbase/salt/_states/container_bridge.py).
+
+```
+cbr0:
+  container_bridge.ensure:
+    - cidr: {{ grains['cbr-cidr'] }}
+    - mtu: 1460
+```
+
```

Comment:
At this point, summarize what you have so far, i.e. what cbr0 is doing in the absence of the stuff that comes later. (namely: bridging between the VM eth0, which corresponds to the VM subnet referred to earlier, and the other end of each container's veth). Also, maybe explain that docker0 is still used for container-to-container traffic (is it?). Also, maybe explain why we can't just hook up docker0 to the VM eth0 and then not need cbr0.